### PR TITLE
Users can join events, Event creators can set max participants

### PR DESCRIPTION
**Pull request by safchow**
## Trello ticket link

[As a user, I want to be able to join events other users have created.](https://trello.com/c/U513rOy9/107-as-a-user-i-want-to-be-able-to-join-events-other-users-have-created)

[As a user, I want to be able to limit the number of users who can attend and RSVP to my events.](https://trello.com/c/WntwvEI7/53-as-a-user-i-want-to-be-able-to-limit-the-number-of-users-who-can-attend-and-rsvp-to-my-events)

## Implementation description
1. General Changes
* Added packages for `@mui/material`
* Updated existing packages to support `@mui/material`
* Removed some unused code leftover in `CreateEvent.js` and `Events.js`
* Removed comments @gabbytang, they looked like possible implementations of code that weren't used. Let me know if they should be restored. 
* Updated the way that the `userID` is obtained through email from Firebase. I noticed some issues with react states being delayed (for example, updating a state then using the state immediately will not reflect the changes made). This caused some issues with requests being sent with null a `userID`.
* Fixed spacing

2. `CreateEvent.js` and `events` table Changes
* Added more columns in the `events` table -> participants, totalParticipants and status
* Added field for participants in "Create Event" page
* Updated `createEvent` API to take the `totalParticipants` value for the event
* Removed the date check from `getEvents` as events that have passed should still be displayed, but their status should be updated respectively (this will need to be implemented in a future PR)

3. `Events.js` Changes
* Created a new table, `eventParticipants` that stores the the participantIDs of the event. 
* Created a new API, `joinEvents`, that allows a user to join a given event, only if the event is not full and the user has not already joined the event.
* Updated the card component for events to include a button where the user can view more details about the event. 
* Added a popup that appears when a user clicks "View Event" that provides more info on the event, and allows users to join the event.
* Added a toast message that informs the user if they've successfully joined the event, have already joined the event, or if the event is full.
* Increased the width of the card component. Note: it is not centred but can be changed later. 

## Steps to test
0. Setup
*  install the new dependencies with `npm i`

1. `CreateEvent.js` Tests
* Verify that the "Event Participants" field throws an error if the form is submitted without the field being filled in.
* Verify that if an event is created and all fields are inputted correctly, the corresponding row appears in the `events` table with the matching `totalParticipants` value from the form.
* Verify that the `participants` value is set to 0 and that the `status` is set to "Active".

2. `Events.js` Tests
* Open up the `events` page and verify the events load correctly.
* Verify that the "View Event" button can be clicked, and the corresponding event's details are displayed in a popup. 
* Verify that the popup can be closed with the "Close" button. 
* Verify that an event can be joined **given that the event has not already been joined**. Verify that a success toast message appears and that the page is refreshed after a short delay. Verify that a corresponding row appears in `eventParticipants`.
* Verify that an event that has already been joined by the user cannot be joined. Verify that an error toast message appears. Verify that a new row **does not** appear in `eventParticipants`.
* Verify that an event at max capacity cannot be joined. Please use the test event, "Max Participant Test" to test this. Verify that an error toast message appears. Verify that a new row **does not** appear in `eventParticipants`. 

## What should reviewers focus on?
* I've added catches for error cases (in the form of toast messages) but if you can think of any ways to break the code that I haven't included in the test steps, please experiment and let me know if anything breaks.

### CODE REVIEW

**Review by gabbytang — CHANGES_REQUESTED** [review]
The event participants field should ideally be number locked. I see there's a catch that sets any entered alphabetic character to "NaN" with an error that blocks the event from being pushed to the db, but once it is set to "NaN" you can't delete/backspace it.
If you misstype or mistakenly type in a letter, the field is locked as null. Submitting the form resets it completely, but it also only does this once. If you put letters in the field a second time, the form is locked as you can't fix the participant field, and the submit button does not nothing.

**Review by gabbytang — CHANGES_REQUESTED** [review]
image
Its up to you whether you want to remove this code I commented out above. Not sure if thats what you were referring to when you mentioned it in the description, but yeah I had included it because I originally was going to show a popup confirmation with the details of the newly created event.

**Review by gabbytang — CHANGES_REQUESTED** [review]
image
I'd probably reword the highlighted message with something simpler like "Event is already full"

**Review by gabbytang — CHANGES_REQUESTED** [review]
Not a huge deal, but maybe we can just remove the "Event" prefix in front of every field label. ie. "Number of Participants" versus "Event Participants" on the Events page.

File: src/components/CreateEvent.js [hunk 1]
```diff
@@ -135,74 +98,92 @@ const CreateEvent = () => {
...
-    } else if (eventDate == '' /*&& enableError == true*/) {
-      console.log("this is the problem");
+    } else if (eventDate === '') {
       setEventDateError(true);
       setEventDateErrorText('Please enter the date of your event in the YYYY-MM_DD format');
       return false
-    } else /*if (enableError == true)*/ {
+    } else if (eventParticipants === '') {
+      setEventParticipantsError(true);
+      setEventParticipantsErrorText('Please enter a valid number for the number of maximum participants of your event.');
```
Comment (gabbytang) [inline src/components/CreateEvent.js]: Maybe change to "Please enter a valid number of maximum participants"

File: src/components/CreateEvent.js [hunk 2]
```diff
@@ -406,6 +385,23 @@ const EventLocation = ({ eventLocation, onEnterEventLocation, eventLocationError
   )
 }
 
+const EventParticipants = ({ eventParticipants, onEnterEventParticipants, eventParticipantsError, eventParticipantsErrorText }) => {
+  return (
+    <Grid item>
+      <TextField
+        id="Participants-of-event"
+        label="Event Participants"
+        placeholder="Enter the maximum number of participants for you event"
```
Comment (gabbytang) [inline src/components/CreateEvent.js]: *your event

**Comment by safchow** [comment]
> The event participants field should ideally be number locked. I see there's a catch that sets any entered alphabetic character to "NaN" with an error that blocks the event from being pushed to the db, but once it is set to "NaN" you can't delete/backspace it. If you misstype or mistakenly type in a letter, the field is locked as null. Submitting the form resets it completely, but it also only does this once. If you put letters in the field a second time, the form is locked as you can't fix the participant field, and the submit button does not nothing.

I've added a check to see if the input is numeric. This should adjust the field to ensure only numbers can be inputted. 
`CreateEvent.js`: 
```
  const handleEventParticipants = (event) => {
    const strToInt = parseInt(event.target.value);

    // If the input is number, update the react state
    if (!isNaN(strToInt)) {
      setEventParticipants(parseInt(event.target.value));
      setEventParticipantsError(false);
      setEventParticipantsErrorText('');
    }

    // Catch where first number can't be deleted, reset the state
    else {
      setEventParticipants('');
      setEventParticipantsError('');
      setEventParticipantsErrorText('');
    }
  }
  ```
 
> Its up to you whether you want to remove this code I commented out above. Not sure if thats what you were referring to when you mentioned it in the description, but yeah I had included it because I originally was going to show a popup confirmation with the details of the newly created event.

I've also added a toast message that appears when an event is submitted successfully. It should disappear after a short delay. If you think it would be better, I can add some details about the event in the message.

**Review by gabbytang — APPROVED** [review]
I think everything looks great. 

There's one thing I would say, but its not really important (more aethetic tbh), but its to include the date of the events on the Events page as default (instead of having to click "VIEW EVENT" to see it. We could change that later or whenever, like I said, not really important. 

You are good to go ahead and merge

**Pull request merged**